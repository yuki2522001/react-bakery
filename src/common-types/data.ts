import { Product } from "./product";

export enum Action {
  GetProductSuccess = "GET_PRODUCT_SUCCESS",
  CreateProductsSuccess = "CREATE_PRODUCT_SUCCESS",
  DeleteProductSuccess = "DELETE_PRODUCT_SUCCESS",
  UpdateProductSuccess = "UPDATE_PRODUCT_SUCCESS",
}

export interface DataState {
  products: Product[];
}

export type DataPayload = Product[] | string | Product

export interface DataAction {
  action: Action;
  payload: DataPayload
}
