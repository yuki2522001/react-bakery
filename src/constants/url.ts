const PRODUCTS_URL = `${process.env.VITE_BASE_URL}/products`;
const CATEGORIES_URL = `${process.env.VITE_BASE_URL}/categories`;

export { PRODUCTS_URL, CATEGORIES_URL };