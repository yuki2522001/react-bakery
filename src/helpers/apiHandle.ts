import axios, { AxiosResponse } from "axios";

// Get data form server
const getData = async <T>(url: string): Promise<T> => {
  const res = await axios.get<T>(url);
  return res.data;
};

// Post new data to server
const create = async <T>(url: string, newData: T):Promise<AxiosResponse<T>> => {
  const res = await axios.post(url, newData);
  return res;
}

// Send a changed data to server using method put to updating the data
const update = async <T>(id: string, dataEdit: T):Promise<AxiosResponse<T>> => {
  const res = await axios.put(id, {...dataEdit});
  return res;
}

// Delete data
const remove = async <T>(id: string): Promise<AxiosResponse<T>> => {
  const res = await axios.delete(id);
  return res;
}

export {
  getData,
  create,
  update,
  remove
}