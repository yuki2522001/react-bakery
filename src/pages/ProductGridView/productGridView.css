.product__list {
  padding: 0 100px;
}

.product__info {
  display: flex;
  flex-wrap: wrap;
}