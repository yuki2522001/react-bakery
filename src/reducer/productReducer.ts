import { Action, DataAction, DataState } from "@common-types/data";
import { Product } from "@common-types/product";

const productReducer = (state: DataState, actions: DataAction): DataState => {
  const { action, payload } = actions;
  switch (action) {
    case Action.GetProductSuccess: {
      return { ...state, products: payload as Product[] };
    }
    case Action.CreateProductsSuccess: {
      return { ...state.products, products: state.products.concat(payload as Product) };
    }
    case Action.DeleteProductSuccess: {
      const deleteProduct = state.products.filter(item => item.id != payload)
      return {...state, products: deleteProduct};
    }
    case Action.UpdateProductSuccess: {
      state.products.findIndex(item => item.id == payload);
      return {...state};
    }
    default: {
      return { ...state };
    }
  }
};

export { productReducer };
