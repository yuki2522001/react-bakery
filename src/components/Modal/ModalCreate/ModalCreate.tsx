import useSWR from "swr";
import React, { ChangeEvent, useCallback, useState } from "react";
import getBase64 from "@helpers/getBase64";
import InputValue from "@components/Input/InputValue/InputValue";
import Button from "@components/common/Button/Button/Button";
import { FieldName, FormProps } from "@common-types/form";
import { Product } from "@common-types/product";
import { CategoryProps } from "@common-types/category";
import { RULES } from "@constants/rules";
import { CATEGORIES_URL } from "@constants/url";
import { getData } from "@helpers/apiHandle";
import { validate } from "@helpers/validate";
import { setFieldsValue } from "@helpers/fieldHandle";
import "../modal.css";

interface ModalCreateProps {
  hideModalCreate: () => void;
  createProduct: (dataProduct: Product) => void;
}

const ModalCreate: React.FC<ModalCreateProps> = ({
  hideModalCreate,
  createProduct,
}) => {
  const initProduct: Product = {
    id: "",
    name: "",
    price: 0,
    images: [],
    quantity: 0,
    categoryId: "",
    description: ""
  }
  const [isDisable, setIsDisable] = useState<boolean>(false);
  const [newProduct, setNewProduct] = useState<Product>(initProduct);
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<Product[]>);
  // create state to handle select file image
  const [selectedFile, setSelectedFile] = useState<string[]>([]);
  // create state to set form values
  const [formValues, setFormValues] = useState<FormProps>({
    categoryId: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    name: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    price: {
      value: "",
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    quantity: {
      value: "",
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    description: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    images: {
      value: "",
      rules: [],
      error: "",
    }
  });

  // handle create product
  const handleCreateProduct = () => {
    const images: string[] = [];
    for (let i = 0; i < selectedFile.length; i++) {
      images.push(selectedFile[i]);
    }
    // validate form
    setFormValues(validate(formValues));
    const temp = (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        if (formValues[fieldName].error) {
          return false;
        }
      }
    );

    // check validate if pass then create product
    if (!temp.includes(false)) {
      setFormValues({...formValues});
      setIsDisable(!isDisable);
      const product: Product = { ...newProduct, images: images }
      createProduct(product);
      hideModalCreate();
    }
  };

  // handle change value
  const handleChange = useCallback((event: { target: { value: string; name: string } }) => {
    const value = event.target.value;
    const fieldName = event.target.name as FieldName;
    setNewProduct({ ...newProduct, [fieldName]: value });
    setFormValues(setFieldsValue(formValues, value, fieldName));
  }, [formValues]);

  // handle change image
  const imageChange = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const files = event.target.files;
    const newFiles = [];
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        newFiles.push(imageSrc)
      }
    }
    setSelectedFile([...selectedFile, ...newFiles]);
  }, [selectedFile]);

  // handle delete image
  const handleDeleteImage = useCallback((event: { target: EventTarget }) => {
    const target = event.target as Element;
    const indexOfArr = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  }, [selectedFile]);

  return (
    <div
      data-testid="modal-create"
      className="modal-create"
      id="bookDeleteModal"
    >
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Add New Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name <span className="star">*</span></label>
              <InputValue
                type="text"
                name="name"
                onChange={handleChange}
              />
              <small className="form__error">
                {formValues?.name?.error ? formValues.name.error : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Description <span className="star">*</span></label>
              <textarea
                data-testid="change-value"
                className="input__text"
                name="description"
                id=""
                cols={30}
                rows={5}
                onChange={handleChange}
              ></textarea>
              <small className="form__error">
                {formValues?.description?.error
                  ? formValues.description.error
                  : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories <span className="star">*</span></label>
              <select
                className="form__select"
                name="categoryId"
                defaultValue={""}
                onChange={handleChange}
              >
                <option value="" disabled>Choose a category ...</option>
                {data?.map(({ id, name }: CategoryProps, index: number) => (
                  <option key={index} value={id}>
                    {name}
                  </option>
                ))}
              </select>
              <small className="form__error">
                {formValues?.categoryId?.error
                  ? formValues.categoryId.error
                  : ""}
              </small>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price <span className="star">*</span></label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  name="price"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues?.price?.error ? formValues.price.error : ""}
                </small>
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity <span className="star">*</span></label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  name="quantity"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues?.quantity?.error ? formValues.quantity.error : ""}
                </small>
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number) => (
                    <img
                      data-testid="after-change-file"
                      className="form__img"
                      key={key}
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <Button
              type="warning"
              text="Cancel"
              onClick={hideModalCreate}
            />
            <Button
              type="success"
              text="Submit"
              disabled = {isDisable}
              onClick={() => handleCreateProduct()}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalCreate;
