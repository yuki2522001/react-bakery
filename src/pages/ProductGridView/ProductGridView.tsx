import useSWR from "swr";
import React, { useCallback, useContext, useEffect, useState } from "react";
import toast from "react-hot-toast";
import ModalCreate from "@components/Modal/ModalCreate/ModalCreate";
import ScrollButton from "@components/common/Button/ScrollButton/ScrollButton";
import ProductGridCard from "@components/ProductGridCard/ProductGridCard";
import { Product } from "@common-types/product";
import { SUCCESS_MSG } from "@constants/message";
import { create, getData, remove } from "@helpers/apiHandle";
import { PRODUCTS_URL } from "@constants/url";
import { DataContext } from "@context/DataContext";
import { Action } from "@common-types/data";
import "./productGridView.css";

const ProductGridView: React.FC = () => {
  const [openModalCreate, setOpenModalCreate] = useState<boolean>(false);
  const { products, searchValue, dispatch } = useContext(DataContext);
  const queryParams: URLSearchParams = new URLSearchParams(searchValue);
  const { data } = useSWR(PRODUCTS_URL + "?" + queryParams.toString(), getData<Product[]>);
  useEffect(() => {
    if(data) {
      dispatch({
        action: Action.GetProductSuccess,
        payload: data
      });
    }
  }, [data]);
  // create product
  const createProduct = async (productData: Product) => {
    const newProduct: Product = {
      id: new Date().valueOf().toString(),
      categoryId: productData.categoryId,
      name: productData.name,
      price: +productData.price,
      quantity: +productData.quantity,
      description: productData.description,
      images: productData.images,
    };
    try {
      const response = await create(PRODUCTS_URL, newProduct);
      if(response) {
        dispatch({
          action: Action.CreateProductsSuccess,
          payload: {...response.data}
        });
        toast.success(SUCCESS_MSG.MESSAGE_ADD_PRODUCT);
      }
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  // delete product
  const deleteProduct = async (id: string) => {
    try {
      const response = await remove(`${PRODUCTS_URL}/${id}`);
      if(response) {
        dispatch({
          action: Action.DeleteProductSuccess,
          payload: id
        });
        toast.success(SUCCESS_MSG.MESSAGE_DELETE_PRODUCT);
      }
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  // handle toggle modal create
  const toggleModalCreate = useCallback(() => {
    setOpenModalCreate(!openModalCreate);
  }, [openModalCreate]);

  return (
    <>
      <div data-testid="product-gird-view" className="product__list">
        <button
          data-testid="open-modal"
          className="btn btn__add"
          onClick={toggleModalCreate}
        >
          Add new product
        </button>
        <div data-testid="delete-product" className="product__info">
          {products?.map((product: Product) => (
            <div className="product__item" key={product.id}>
              <ProductGridCard
                product={product}
                deleteProduct={deleteProduct}
              />
            </div>
          ))}
        </div>
        {openModalCreate && (
          <ModalCreate
            createProduct={createProduct}
            hideModalCreate={toggleModalCreate}
          />
        )}
      </div>
      <ScrollButton className="btn__backToTop" children={<i className="fa fa-arrow-alt-circle-up"></i>} />
    </>
  );
};

export default ProductGridView;
