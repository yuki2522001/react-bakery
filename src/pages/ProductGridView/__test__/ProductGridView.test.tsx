import "@testing-library/jest-dom";
import mockAxios from "@__mocks__/axios";
import ProductGridView from "../ProductGridView";
import ModalCreate from "@components/Modal/ModalCreate/ModalCreate";
import { fireEvent, render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { PRODUCTS_URL } from "@constants/url";
import { create, getData, remove } from "@helpers/apiHandle";
import { useState } from "react";
import {
  PRODUCT_MOCKING,
  PRODUCT_MOCKING_LIST,
} from "@__mocks__/constants/product";
import { Action, DataState } from "@common-types/data";
import { productReducer } from "@reducer/productReducer";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Product grid view component", () => {
  beforeEach(() => {
    (useState as jest.Mock).mockImplementation(
      jest.requireActual("react").useState
    );
  });

  afterEach(() => {
    mockAxios.reset();
  });

  const setup = () => {
    const utils = render(
      <ModalCreate hideModalCreate={() => {}} createProduct={() => {}} />
    );
    const input = utils.getByTestId("change-value") as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };

  test("should change value when onChang input", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: "Cheese pocket" } });
    expect(input.value).toBe("Cheese pocket");
  });

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: PRODUCT_MOCKING_LIST });
    const result = await getData(PRODUCTS_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(PRODUCTS_URL);
    expect(result).toEqual(PRODUCT_MOCKING_LIST);
  });

  test("add new product item should call", async () => {
    mockAxios.post.mockResolvedValueOnce(PRODUCT_MOCKING);
    const result = await create(PRODUCTS_URL, PRODUCT_MOCKING);
    expect(mockAxios.post).toHaveBeenCalledWith(PRODUCTS_URL, PRODUCT_MOCKING);
    expect(result).toEqual(PRODUCT_MOCKING);
  });

  test("delete product item should call", async () => {
    const PRODUCT_URL_CALL = PRODUCTS_URL + "/1";
    mockAxios.delete.mockResolvedValueOnce(PRODUCT_MOCKING);
    const result = await remove(PRODUCT_URL_CALL);
    expect(mockAxios.delete).toHaveBeenCalledWith(PRODUCT_URL_CALL);
    expect(result).toEqual(PRODUCT_MOCKING);
  });

  test("should render product grid view component", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <ProductGridView />
      </Router>
    );
    expect(getByTestId("product-gird-view")).toBeInTheDocument();
  });

  test("should open modal when click button 'Add new product'", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <ProductGridView />
      </Router>
    );
    const btnOpenModal = getByTestId("open-modal");
    fireEvent.click(btnOpenModal);
    expect(btnOpenModal).toBeInTheDocument();
  });

  test("should create product when dispatch action CreateProductsSuccess", () => {
    const initialState: DataState = {
      products: PRODUCT_MOCKING_LIST,
    };
    const createProduct = {
      action: Action.CreateProductsSuccess,
      payload: PRODUCT_MOCKING_LIST,
    };
    const updatedState = productReducer(initialState, createProduct);
    expect(updatedState).toEqual(updatedState);
  });

  test("should get data when dispatch action GetProductSuccess", () => {
    const initialState: DataState = {
      products: PRODUCT_MOCKING_LIST,
    };
    const createProduct = {
      action: Action.GetProductSuccess,
      payload: PRODUCT_MOCKING_LIST,
    };
    const updatedState = productReducer(initialState, createProduct);
    expect(updatedState).toEqual(updatedState);
  });

  test("should delete product when dispatch action DeleteProductSuccess", () => {
    const initialState: DataState = {
      products: PRODUCT_MOCKING_LIST,
    };
    const deleteProduct = {
      action: Action.DeleteProductSuccess,
      payload: PRODUCT_MOCKING_LIST,
    };
    const updatedState = productReducer(initialState, deleteProduct);
    expect(updatedState).toEqual(updatedState);
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Router location={history.location} navigator={history}>
        <ProductGridView />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
