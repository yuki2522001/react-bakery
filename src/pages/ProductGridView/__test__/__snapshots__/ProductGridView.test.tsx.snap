// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product grid view component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="product__list"
    data-testid="product-gird-view"
  >
    <button
      class="btn btn__add"
      data-testid="open-modal"
    >
      Add new product
    </button>
    <div
      class="product__info"
      data-testid="delete-product"
    />
  </div>
  <button
    class="btn__backToTop"
    data-testid="back-to-top"
  >
    <i
      class="fa fa-arrow-alt-circle-up"
    />
  </button>
  <div />
</DocumentFragment>
`;
